### PR TITLE
Ea/data lab

### PR DESCRIPTION
**Pull request by eanders**

### CODE REVIEW

**Comment by eanders** [comment]
@awisema, if you can review this, that'd be great.  There's one failing test that passes locally.  Need to see if I can run the entire test suite locally and duplicate the issue to fix before we merge.

File: app/models/grda_warehouse/hud/enrollment.rb [hunk 1]
```diff
@@ -55,23 +55,23 @@ class Enrollment < Base
 
     # Income benefits at various stages
     has_one :income_benefits_at_entry, -> do
-      at_entry
+      at_entry.joins(:enrollment).where(ib_t[:InformationDate].eq(e_t[:EntryDate]))
```
Comment (awisema) [inline app/models/grda_warehouse/hud/enrollment.rb:59]: You rolled the date match into `at_entry`/`at_exit` , so these are redundant?

File: app/models/grda_warehouse/hud/income_benefit.rb [hunk 2]
```diff
@@ -37,11 +39,11 @@ class IncomeBenefit < Base
     }
 
     scope :at_entry, -> do
-      where(DataCollectionStage: 1)
+      where(DataCollectionStage: 1).joins(:enrollment).where(ib_t[:InformationDate].eq(e_t[:EntryDate]))
```
Comment (awisema) [inline app/models/grda_warehouse/hud/income_benefit.rb:43]: Are these ever used for DQ? In which case, do you need to add something like `-> (on_date: true)`?

File: drivers/hmis_csv_twenty_twenty_two/spec/models/exporter/multi_enrollment_tests.rb [hunk 3]
```diff
@@ -79,12 +79,10 @@ def involved_clients
           end.map(&:id).
             sort. # Sort by number, because export chose first 3 enrollments and related items before converting to strings
             map(&:to_s)
-          test_data = instance_variable_get("@#{items}").select do |m|
+          instance_variable_get("@#{items}").select do |m|
             involved_enrollment_project_entry_ids.include? m.EnrollmentID
           end.map { |m| [m.id, m.EnrollmentID, m.data_source_id] }
```
Comment (awisema) [inline drivers/hmis_csv_twenty_twenty_two/spec/models/exporter/multi_enrollment_tests.rb:84]: Dead code?

File: drivers/hud_apr/app/models/hud_apr/generators/shared/fy2021/base.rb [hunk 4]
```diff
@@ -114,6 +114,9 @@ def needs_ce_assessments?
           exit_record = last_service_history_enrollment.enrollment if exit_date.present? && exit_date <= @report.end_date
 
           income_at_start = enrollment.income_benefits_at_entry
+          # Only include income at start if the information date matches entry date
```
Comment (awisema) [inline drivers/hud_apr/app/models/hud_apr/generators/shared/fy2021/base.rb]: Done in `income_benefits_at_entry`?

File: drivers/hud_spm_report/app/models/hud_spm_report/generators/fy2020/base.rb [hunk 5]
```diff
@@ -1428,30 +1465,115 @@ def hoh_destination_for(project_types, client_id, household_id)
...
+      # group by date so we can reject dates where stop-projects are present
+      all_nights = all_nights.group_by { |night| night[:date] }
+      # remove any days with a stop project type
+      all_nights.each do |date, nights|
+        if line.in?([:m1a1, :m1b1])
+          remove_homeless_if = nights_for_negation[date].any? { |night| night[:project_type].in?(TH) || (night[:project_type].in?(PH) && night[:housed]) }
+          nights.reject! { |night| night[:project_type].in?(ES_SH) } if remove_homeless_if
+          # NOTE: the spec says:
+          # a. For measures 1a.1 and 1b.1, time spent by clients housed in TH or PH projects negates overlapping time spent in ES and SH projects.
+          # data labe test kit seems to reject any homeless time
```
Comment (awisema) [inline drivers/hud_spm_report/app/models/hud_spm_report/generators/fy2020/base.rb]: I think the AAQ confirmed this behavior as correct? Update this to reflect the referenced explanation?

File: drivers/hud_spm_report/app/models/hud_spm_report/generators/fy2020/base.rb [hunk 6]
```diff
@@ -312,12 +328,26 @@ def she_household_column
...
-        )
+        ).uniq
 
         # transform them into per client metrics
         pending_associations = nights_for_batch.group_by do |r|
           r.fetch(:client_id)
         end.map do |client_id, nights|
+          client = clients_by_id.fetch(client_id)
+          # debug = "#{client.first_name} #{client.last_name}" == "Amber Majestic"
+          # binding.pry if debug
```
Comment (awisema) [inline drivers/hud_spm_report/app/models/hud_spm_report/generators/fy2020/base.rb]:
```suggestion
```

File: spec/rails_helper.rb [hunk 7]
```diff
@@ -70,7 +70,7 @@
 
   config.before(:suite) do
     Dir.glob('{drivers,spec}/**/fixpoints/*.yml').each do |filename|
-      FileUtils.rm(filename)
+      # FileUtils.rm(filename)
```
Comment (awisema) [inline spec/rails_helper.rb]:
```suggestion
      # This cleans up the fixpoints before each rspec run
      # Disabling the next line, can make iterative test development faster.
      FileUtils.rm(filename)
```

**Review by awisema — CHANGES_REQUESTED** [review]
Some small clean up suggestions; it is such a relief to see this over the finishline!

**Review by awisema — APPROVED** [review]
